File: HEAP/MaxHeap.cpp
#include "vector"
#include "iostream"
using namespace std;

 
class MaxHeap{
private:
	vector<int> items;

public:
	int size();
	void softDown();
	void softUp();
	void insert(const int& item);
	int peekMax();
	void removeMax();
	void toString();
	//void toString_Queue();
	bool isEmpty();

};

bool MaxHeap::isEmpty(){
	return items.empty();
}

void MaxHeap::toString(){
	for (int i = 0; i < items.size(); i++){
		cout << items[i] << " ";
	}
	cout << endl;
}


int MaxHeap::size(){
	return items.size();
}

 
void MaxHeap::softUp(){
	int k = items.size() - 1;
	while (k > 0){
		int p = (k - 1) / 2;
		int child = items[k];
		int parent = items[p];
		if (child > parent){
			items[k] = parent;
			items[p] = child;
			k = p;
		}else break;
	}
}

 
void MaxHeap::softDown(){
	int parent = 0;
	int child1 = parent * 2 + 1;

	while (child1 < items.size()){
		int max = child1;
		int child2 = child1 + 1;
		if (child2 < items.size()){
			if (items[child2] > items[child1]){
				max = child2;
			}
		}
		
		if (items[parent] < items[max]){
			int temp = items[parent];
			items[parent] = items[max];
			items[max] = temp;
			parent = max;
			child1 = 2 * parent + 1;
		}else break;
	}
}

 
void MaxHeap::insert(const int& item){
	items.push_back(item);
	softUp();
}

 
int MaxHeap::peekMax(){
	return items[0];
}

 
void MaxHeap::removeMax(){
	if (items.size() == 0){
		throw new exception();
	}
	items[0] = items.back();
	items.pop_back();
	softDown();
}


int main(){

	MaxHeap maxHeap;
	for (int i = 0; i <= 10; i++){
		maxHeap.insert(i);
	}
	maxHeap.removeMax();
	cout << maxHeap.peekMax() << endl;
	maxHeap.toString();
	cout << maxHeap.isEmpty() << endl;
	/*for (int i = 0; i <= 9; i++){
		maxHeap.removeMax();
	}
	cout << maxHeap.isEmpty() << endl;*/
	while (!maxHeap.isEmpty()){
		cout << maxHeap.peekMax() << endl;
		maxHeap.removeMax();
	}
		
}

File: HEAP/MaxHeap.java

 
import java.util.ArrayList;
import java.util.NoSuchElementException;
import java.util.Scanner;
 
public class MaxHeap<T extends Comparable<T>> {
     
    private ArrayList<T> items;
     
    public MaxHeap() {
        items = new ArrayList<T>();
    }
    
    public static void main(String[] args){
    	
    	MaxHeap<Integer> hp1 = new MaxHeap<Integer>();
    	
    	
    	for(int i = 0;i <= 10;i++){
    		hp1.insert(i);
    	}
    	hp1.removeMax();
    	System.out.println(hp1);
    	
    	
    	Scanner sc = new Scanner(System.in);
    	System.out.print("Enter next int,'done' to stop :");;
    	String line = sc.next();
    	
    	
    	while(!line.equals("done")){
    		hp1.insert(Integer.parseInt(line));
    		System.out.println(hp1);
    		System.out.print("enter next int,'done' to stop");
	    		line = sc.next();
	    	}
    	
    	while(!hp1.isEmpty()){
    		int max = hp1.removeMax();
    		System.out.println(max + " " + hp1);
    	}
    	sc.close();
    }
    	
    
    private void softUp() {
        int k = items.size() - 1;
        while (k > 0) {
            int p = (k-1)/2;
            T item = items.get(k);
            T parent = items.get(p);
            if (item.compareTo(parent) > 0) {
                // swap
                items.set(k, parent);
                items.set(p, item);
                 
                // move up one level
                k = p;
            } else {
                break;
            }
        }
    }
     
    public void insert(T item) {
        items.add(item);
        softUp();
    }
     
    private void softDown() {
        int k = 0;
        int l = 2*k+1;
        while (l < items.size()) {
            int max=l, r=l+1;
            if (r < items.size()) {
                if (items.get(r).compareTo(items.get(l)) > 0) {
                    max++;
                }
            }
            if (items.get(k).compareTo(items.get(max)) < 0) {
                    // switch
                    T temp = items.get(k);
                    items.set(k, items.get(max));
                    items.set(max, temp);
                    k = max;
                    l = 2*k+1;
            } else {
                break;
            }
        }
    }
     
    public T removeMax() 
    throws NoSuchElementException {
        if (items.size() == 0) {
            throw new NoSuchElementException();
        }
        if (items.size() == 1) {
            return items.remove(0);
        }
        T hold = items.get(0);
        items.set(0, items.remove(items.size()-1));
        softDown();
        return hold;
    }
 
    public int size() {
        return items.size();
    }
     
    public boolean isEmpty() {
        return items.isEmpty();
         
    }
     
    public String toString() {
        return items.toString();
    }
}